src/metric/rank_metric.cc

```diff
@@ -451,82 +465,88 @@ struct EvalAucPR : public Metric {
-                 bool distributed) {
-    CHECK_NE(info.labels_.Size(), 0U) << "label set cannot be empty";
-    CHECK_EQ(preds.Size(), info.labels_.Size())
-        << "label size predict size not match";
-    std::vector<unsigned> tgptr(2, 0);
-    tgptr[1] = static_cast<unsigned>(info.labels_.Size());
-    const std::vector<unsigned> &gptr =
-        info.group_ptr_.size() == 0 ? tgptr : info.group_ptr_;
-    CHECK_EQ(gptr.back(), info.labels_.Size())
-        << "EvalAucPR: group structure must match number of prediction";
-    const auto ngroup = static_cast<bst_omp_uint>(gptr.size() - 1);
+                 bool distributed,
+                 const std::vector<unsigned> &gptr) {
+    const auto ngroups = static_cast<bst_omp_uint>(gptr.size() - 1);
+
     // sum of all AUC's across all query groups
     double sum_auc = 0.0;
     int auc_error = 0;
-    // each thread takes a local rec
-    std::vector<std::pair<bst_float, unsigned>> rec;
-    const auto& h_labels = info.labels_.HostVector();
-    const std::vector<bst_float>& h_preds = preds.HostVector();
-
-    for (bst_omp_uint group_id = 0; group_id < ngroup; ++group_id) {
-      double total_pos = 0.0;
-      double total_neg = 0.0;
-      rec.clear();
-      for (unsigned j = gptr[group_id]; j < gptr[group_id + 1]; ++j) {
-        const bst_float wt
-          = WeightPolicy::GetWeightOfInstance(info, j, group_id);
-        total_pos += wt * h_labels[j];
-        total_neg += wt * (1.0f - h_labels[j]);
-        rec.emplace_back(h_preds[j], j);
-      }
-      XGBOOST_PARALLEL_STABLE_SORT(rec.begin(), rec.end(), common::CmpFirst);
-      // we need pos > 0 && neg > 0
-      if (0.0 == total_pos || 0.0 == total_neg) {
-        auc_error += 1;
-        continue;
-      }
-      // calculate AUC
-      double tp = 0.0, prevtp = 0.0, fp = 0.0, prevfp = 0.0, h = 0.0, a = 0.0, b = 0.0;
-      for (size_t j = 0; j < rec.size(); ++j) {
-        const bst_float wt
-          = WeightPolicy::GetWeightOfSortedRecord(info, rec, j, group_id);
-        tp += wt * h_labels[rec[j].second];
-        fp += wt * (1.0f - h_labels[rec[j].second]);
-        if ((j < rec.size() - 1 && rec[j].first != rec[j + 1].first) || j  == rec.size() - 1) {
-          if (tp == prevtp) {
-            a = 1.0;
-            b = 0.0;
-          } else {
-            h = (fp - prevfp) / (tp - prevtp);
-            a = 1.0 + h;
-            b = (prevfp - h * prevtp) / total_pos;
-          }
-          if (0.0 != b) {
-            sum_auc += (tp / total_pos - prevtp / total_pos -
-                        b / a * (std::log(a * tp / total_pos + b) -
-                                 std::log(a * prevtp / total_pos + b))) / a;
-          } else {
-            sum_auc += (tp / total_pos - prevtp / total_pos) / a;
+
+    const auto &h_labels = info.labels_.ConstHostVector();
+    const auto &h_preds = preds.ConstHostVector();
+
+    #pragma omp parallel reduction(+:sum_auc, auc_error) if (ngroups > 1)
```

Comment:
I spent some time working on the deterministic algorithm recently: https://github.com/dmlc/xgboost/pull/5361 .  I would like to know whether a component is deterministic or not.